NeutrinoClient data race

## Problem
The `NeturinoClient` `Rescan` method locks, sets `rescan` to `nil` then unlocks itself before resetting the value.
As the client state is `scanning = true` yet `rescan = nil` when the lock is released there is a race between the executing `Rescan` method and any `NotifyReceived` calls that are waiting to obtain the lock.  If a `NotifyReceived` call picks up the lock in this state, it will panic when it tries to run `rescan.Update` as `scanning` is `true`.

https://github.com/btcsuite/btcwallet/blob/master/chain/neutrino.go#L351

__Originally analyzed and found by @guggero in https://github.com/lightningnetwork/lnd/issues/6985__


```
--- FAIL: TestLightningWallet (7.74s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x7d47b7]

goroutine 6 [running]:
testing.tRunner.func1.2({0x9c34e0, 0xfdbfe0})
	/opt/hostedtoolcache/go/1.19.0/x64/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.19.0/x64/src/testing/testing.go:1399 +0x39f
panic({0x9c34e0, 0xfdbfe0})
	/opt/hostedtoolcache/go/1.19.0/x64/src/runtime/panic.go:884 +0x212
github.com/lightninglabs/neutrino.(*Rescan).Update(0x0, {0xc00064a248, 0x1, 0xc00129a4e0?})
	/home/runner/work/go/pkg/mod/github.com/lightninglabs/neutrino@v0.14.2/rescan.go:1391 +0x57
github.com/btcsuite/btcwallet/chain.(*NeutrinoClient).NotifyReceived(0xc0000fa120, {0xc0012bc060?, 0x1, 0x1})
	/home/runner/work/go/pkg/mod/github.com/btcsuite/btcwallet@v0.16.1/chain/neutrino.go:457 +0x159
github.com/btcsuite/btcwallet/wallet.(*Wallet).NewAddress(0xc0001ce480, 0x0, {0x0?, 0x9f16e0?})
	/home/runner/work/go/pkg/mod/github.com/btcsuite/btcwallet@v0.16.1/wallet/wallet.go:3020 +0x17a
github.com/lightningnetwork/lnd/lnwallet/btcwallet.(*BtcWallet).NewAddress(0xc0005a8180, 0x50?, 0x0, {0xa88a60?, 0x7?})
	/home/runner/work/lnd/lnd/lnwallet/btcwallet/btcwallet.go:556 +0x10c
github.com/lightningnetwork/lnd/lnwallet/test.loadTestCredits(0xc000078720, 0xc000340380, 0x14, 0x0?)
	/home/runner/work/lnd/lnd/lnwallet/test/test_interface.go:275 +0x287
github.com/lightningnetwork/lnd/lnwallet/test.createTestWallet({0xc00031ff80?, 0xa87ad1?}, 0x5?, 0xfe8080, {0xc43560, 0xc0001ca180}, {0xc48820, 0xc0005a8180}, {0xc43b00, 0xc0011c41c8}, ...)
	/home/runner/work/lnd/lnd/lnwallet/test/test_interface.go:378 +0x4a9
github.com/lightningnetwork/lnd/lnwallet/test.runTests(_, _, {_, _}, _, {{0xc00002eba0, 0xe}, {0xa873bc, 0x2}, {0xa879c6, ...}, ...}, ...)
	/home/runner/work/lnd/lnd/lnwallet/test/test_interface.go:3398 +0x2633
github.com/lightningnetwork/lnd/lnwallet/test.TestLightningWallet(0xc000007ba0, {0xa89361, 0x8})
	/home/runner/work/lnd/lnd/lnwallet/test/test_interface.go:3064 +0x8ef
github.com/lightningnetwork/lnd/lnwallet/test/neutrino_test.TestLightningWallet(0x0?)
	/home/runner/work/lnd/lnd/lnwallet/test/neutrino/neutrino_test.go:12 +0x25
testing.tRunner(0xc000007ba0, 0xb9f9b8)
	/opt/hostedtoolcache/go/1.19.0/x64/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.19.0/x64/src/testing/testing.go:1493 +0x35f
FAIL	github.com/lightningnetwork/lnd/lnwallet/test/neutrino	7.770s
FAIL
```

### Related Work

See PR #695
See PR #533

## Solution 1
Make a change against `btcwallet` to address the state management issues:
1. make the entire `Rescan` method a critical path (i.e., only unlock via a single defer call)
2. make the entire `NotifyReceived` method a critical path
3. remove the locking in the `notificationHandler` to prevent a deadlock caused by making `Rescan` completely critical.  This locking isn't necessary as it attempts to read the state of the `rescanErr` channel.  But if `rescanErr` is nil, then there is no `Rescan` object created yet so no errors to read.  Also, reading from a `nil` channel will simply block but the read is wrapped in a select statement with other exits.

### Risks
- there are no unit tests on the neutrino client in btcwallet and even a targeted change like this is hard to know if it introduced any regressions

--- a/.gitignore
+++ b/.gitignore
@@ -3,3 +3,4 @@ vendor
 .idea
 coverage.txt
 *.swp
+.vscode

--- a/chain/chainservice.go
+++ b/chain/chainservice.go
@@ -1,0 +1,40 @@
+package chain
+
+import (
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/btcutil/gcs"
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/neutrino"
+	"github.com/lightninglabs/neutrino/banman"
+	"github.com/lightninglabs/neutrino/headerfs"
+)
+
+// NeutrinoChainService is an interface that encapsulates all the public
+// methods of a *neutrino.ChainService
+type NeutrinoChainService interface {
+	Start() error
+	GetBlock(chainhash.Hash, ...neutrino.QueryOption) (*btcutil.Block, error)
+	GetBlockHeight(*chainhash.Hash) (int32, error)
+	BestBlock() (*headerfs.BlockStamp, error)
+	GetBlockHash(int64) (*chainhash.Hash, error)
+	GetBlockHeader(*chainhash.Hash) (*wire.BlockHeader, error)
+	IsCurrent() bool
+	SendTransaction(*wire.MsgTx) error
+	GetCFilter(chainhash.Hash, wire.FilterType,
+		...neutrino.QueryOption) (*gcs.Filter, error)
+	GetUtxo(...neutrino.RescanOption) (*neutrino.SpendReport, error)
+	BanPeer(string, banman.Reason) error
+	IsBanned(addr string) bool
+	AddPeer(*neutrino.ServerPeer)
+	AddBytesSent(uint64)
+	AddBytesReceived(uint64)
+	NetTotals() (uint64, uint64)
+	UpdatePeerHeights(*chainhash.Hash, int32, *neutrino.ServerPeer)
+	ChainParams() chaincfg.Params
+	Stop() error
+	PeerByAddr(string) *neutrino.ServerPeer
+}
+
+var _ NeutrinoChainService = (*neutrino.ChainService)(nil)

--- a/chain/mocks_test.go
+++ b/chain/mocks_test.go
@@ -1,0 +1,157 @@
+package chain
+
+import (
+	"container/list"
+	"errors"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/btcutil/gcs"
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/neutrino"
+	"github.com/lightninglabs/neutrino/banman"
+	"github.com/lightninglabs/neutrino/headerfs"
+)
+
+var (
+	errNotImplemented = errors.New("not implemented")
+	testBestBlock     = &headerfs.BlockStamp{
+		Height: 42,
+	}
+)
+
+var (
+	_ rescanner            = (*mockRescanner)(nil)
+	_ NeutrinoChainService = (*mockChainService)(nil)
+)
+
+// newMockNeutrinoClient constructs a neutrino client with a mock chain
+// service implementation and mock rescanner interface implementation.
+func newMockNeutrinoClient() *NeutrinoClient {
+	// newRescanFunc returns a mockRescanner
+	newRescanFunc := func(ro ...neutrino.RescanOption) rescanner {
+		return &mockRescanner{
+			updateArgs: list.New(),
+		}
+	}
+
+	return &NeutrinoClient{
+		CS:        &mockChainService{},
+		newRescan: newRescanFunc,
+	}
+}
+
+// mockRescanner is a mock implementation of a rescanner interface for use in
+// tests.  Only the Update method is implemented.
+type mockRescanner struct {
+	updateArgs *list.List
+}
+
+func (m *mockRescanner) Update(opts ...neutrino.UpdateOption) error {
+	m.updateArgs.PushBack(opts)
+	return nil
+}
+
+func (m *mockRescanner) Start() <-chan error {
+	return nil
+}
+
+func (m *mockRescanner) WaitForShutdown() {
+	// no-op
+}
+
+// mockChainService is a mock implementation of a chain service for use in
+// tests.  Only the Start, GetBlockHeader and BestBlock methods are implemented.
+type mockChainService struct {
+}
+
+func (m *mockChainService) Start() error {
+	return nil
+}
+
+func (m *mockChainService) BestBlock() (*headerfs.BlockStamp, error) {
+	return testBestBlock, nil
+}
+
+func (m *mockChainService) GetBlockHeader(
+	*chainhash.Hash) (*wire.BlockHeader, error) {
+
+	return &wire.BlockHeader{}, nil
+}
+
+func (m *mockChainService) GetBlock(chainhash.Hash,
+	...neutrino.QueryOption) (*btcutil.Block, error) {
+
+	return nil, errNotImplemented
+}
+
+func (m *mockChainService) GetBlockHeight(*chainhash.Hash) (int32, error) {
+	return 0, errNotImplemented
+}
+
+func (m *mockChainService) GetBlockHash(int64) (*chainhash.Hash, error) {
+	return nil, errNotImplemented
+}
+
+func (m *mockChainService) IsCurrent() bool {
+	return false
+}
+
+func (m *mockChainService) SendTransaction(*wire.MsgTx) error {
+	return errNotImplemented
+}
+
+func (m *mockChainService) GetCFilter(chainhash.Hash,
+	wire.FilterType, ...neutrino.QueryOption) (*gcs.Filter, error) {
+
+	return nil, errNotImplemented
+}
+
+func (m *mockChainService) GetUtxo(
+	_ ...neutrino.RescanOption) (*neutrino.SpendReport, error) {
+
+	return nil, errNotImplemented
+}
+
+func (m *mockChainService) BanPeer(string, banman.Reason) error {
+	return errNotImplemented
+}
+
+func (m *mockChainService) IsBanned(addr string) bool {
+	panic(errNotImplemented)
+}
+
+func (m *mockChainService) AddPeer(*neutrino.ServerPeer) {
+	panic(errNotImplemented)
+}
+
+func (m *mockChainService) AddBytesSent(uint64) {
+	panic(errNotImplemented)
+}
+
+func (m *mockChainService) AddBytesReceived(uint64) {
+	panic(errNotImplemented)
+}
+
+func (m *mockChainService) NetTotals() (uint64, uint64) {
+	panic(errNotImplemented)
+}
+
+func (m *mockChainService) UpdatePeerHeights(*chainhash.Hash,
+	int32, *neutrino.ServerPeer,
+) {
+	panic(errNotImplemented)
+}
+
+func (m *mockChainService) ChainParams() chaincfg.Params {
+	panic(errNotImplemented)
+}
+
+func (m *mockChainService) Stop() error {
+	panic(errNotImplemented)
+}
+
+func (m *mockChainService) PeerByAddr(string) *neutrino.ServerPeer {
+	panic(errNotImplemented)
+}

--- a/chain/neutrino.go
+++ b/chain/neutrino.go
@@ -20,9 +20,9 @@ import (
 	"github.com/lightninglabs/neutrino/headerfs"
 )
 
-// NeutrinoClient is an implementation of the btcwalet chain.Interface interface.
+// NeutrinoClient is an implementation of the btcwallet chain.Interface interface.
 type NeutrinoClient struct {
-	CS *neutrino.ChainService
+	CS NeutrinoChainService
 
 	chainParams *chaincfg.Params
 

--- a/chain/neutrino.go
+++ b/chain/neutrino.go
@@ -27,8 +27,14 @@ type NeutrinoClient struct {
 	chainParams *chaincfg.Params
 
 	// We currently support only one rescan/notification goroutine per client.
+	// Therefore there can only be one instance of the rescan object and
+	// the rescanMtx synchronizes its access.  Calls to the NotifyReceived
+	// and Rescan methods of the client must hold the rescan mutex lock for
+	// the length of their execution to ensure that all operations that
+	// affect the rescan object are atomic.
 	rescan    rescanner
 	newRescan newRescanFunc
+	rescanMtx sync.Mutex
 
 	enqueueNotification     chan interface{}
 	dequeueNotification     chan interface{}
@@ -46,6 +52,14 @@ type NeutrinoClient struct {
 	finished   bool
 	isRescan   bool
 
+	// The clientMtx synchronizes access to the state variables of the client.
+	//
+	// TODO(mstreet3): Currently the clientMtx synchronizes access to the
+	// rescanQuit and rescanErr channels, which cancel the current rescan
+	// goroutine when closed and is updated each time a new rescan goroutine
+	// is created, respectively.  All state related to the rescan goroutine
+	// should ideally be synchronized by the same lock or via some other
+	// shared mechanism.
 	clientMtx sync.Mutex
 }
 
@@ -351,6 +365,10 @@ func (s *NeutrinoClient) pollCFilter(hash *chainhash.Hash) (*gcs.Filter, error) 
 func (s *NeutrinoClient) Rescan(startHash *chainhash.Hash, addrs []btcutil.Address,
 	outPoints map[wire.OutPoint]btcutil.Address) error {
 
+	// Obtain and hold the rescan mutex lock for the duration of the call.
+	s.rescanMtx.Lock()
+	defer s.rescanMtx.Unlock()
+
 	s.clientMtx.Lock()
 	if !s.started {
 		s.clientMtx.Unlock()
@@ -458,6 +476,10 @@ func (s *NeutrinoClient) NotifyBlocks() error {
 
 // NotifyReceived replicates the RPC client's NotifyReceived command.
 func (s *NeutrinoClient) NotifyReceived(addrs []btcutil.Address) error {
+	// Obtain and hold the rescan mutex lock for the duration of the call.
+	s.rescanMtx.Lock()
+	defer s.rescanMtx.Unlock()
+
 	s.clientMtx.Lock()
 
 	// If we have a rescan running, we just need to add the appropriate

--- a/chain/neutrino.go
+++ b/chain/neutrino.go
@@ -100,24 +100,36 @@ func (s *NeutrinoClient) Start() error {
 	s.clientMtx.Lock()
 	defer s.clientMtx.Unlock()
 	if !s.started {
+		// Reset the client state.
 		s.enqueueNotification = make(chan interface{})
 		s.dequeueNotification = make(chan interface{})
 		s.currentBlock = make(chan *waddrmgr.BlockStamp)
 		s.quit = make(chan struct{})
 		s.started = true
+
+		// Go place a ClientConnected notification onto the queue.
 		s.wg.Add(1)
 		go func() {
+			defer s.wg.Done()
+
 			select {
 			case s.enqueueNotification <- ClientConnected{}:
 			case <-s.quit:
 			}
 		}()
+
+		// Go launch the notification handler.
+		s.wg.Add(1)
 		go s.notificationHandler()
 	}
 	return nil
 }
 
 // Stop replicates the RPC client's Stop method.
+//
+// TODO(mstreet3): The Stop method does not cancel the long-running rescan
+// goroutine.  This is a memory leak.  Stop should shutdown the rescan goroutine
+// and reset the scanning state of the NeutrinoClient to false.
 func (s *NeutrinoClient) Stop() {
 	s.clientMtx.Lock()
 	defer s.clientMtx.Unlock()

--- a/chain/neutrino.go
+++ b/chain/neutrino.go
@@ -26,8 +26,9 @@ type NeutrinoClient struct {
 
 	chainParams *chaincfg.Params
 
-	// We currently support one rescan/notifiction goroutine per client
-	rescan *neutrino.Rescan
+	// We currently support only one rescan/notification goroutine per client.
+	rescan    rescanner
+	newRescan newRescanFunc
 
 	enqueueNotification     chan interface{}
 	dequeueNotification     chan interface{}
@@ -53,9 +54,21 @@ type NeutrinoClient struct {
 func NewNeutrinoClient(chainParams *chaincfg.Params,
 	chainService *neutrino.ChainService) *NeutrinoClient {
 
+	chainSource := &neutrino.RescanChainSource{
+		ChainService: chainService,
+	}
+
+	// Adapt the neutrino.NewRescan constructor to satisfy the
+	// newRescanFunc type by closing over the chainSource and
+	// passing in the rescan options.
+	newRescan := func(ropts ...neutrino.RescanOption) rescanner {
+		return neutrino.NewRescan(chainSource, ropts...)
+	}
+
 	return &NeutrinoClient{
 		CS:          chainService,
 		chainParams: chainParams,
+		newRescan:   newRescan,
 	}
 }
 
@@ -411,10 +424,7 @@ func (s *NeutrinoClient) Rescan(startHash *chainhash.Hash, addrs []btcutil.Addre
 	}
 
 	s.clientMtx.Lock()
-	newRescan := neutrino.NewRescan(
-		&neutrino.RescanChainSource{
-			ChainService: s.CS,
-		},
+	newRescan := s.newRescan(
 		neutrino.NotificationHandlers(rpcclient.NotificationHandlers{
 			OnBlockConnected:         s.onBlockConnected,
 			OnFilteredBlockConnected: s.onFilteredBlockConnected,
@@ -466,10 +476,7 @@ func (s *NeutrinoClient) NotifyReceived(addrs []btcutil.Address) error {
 	s.lastFilteredBlockHeader = nil
 
 	// Rescan with just the specified addresses.
-	newRescan := neutrino.NewRescan(
-		&neutrino.RescanChainSource{
-			ChainService: s.CS,
-		},
+	newRescan := s.newRescan(
 		neutrino.NotificationHandlers(rpcclient.NotificationHandlers{
 			OnBlockConnected:         s.onBlockConnected,
 			OnFilteredBlockConnected: s.onFilteredBlockConnected,

--- a/chain/neutrino_test.go
+++ b/chain/neutrino_test.go
@@ -1,0 +1,218 @@
+package chain
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// maxDur is the max duration a test has to execute successfully.
+var maxDur = 5 * time.Second
+
+// TestNeutrinoClientSequentialStartStop ensures that the client
+// can sequentially Start and Stop without errors or races.
+func TestNeutrinoClientSequentialStartStop(t *testing.T) {
+	var (
+		nc           = newMockNeutrinoClient()
+		wantRestarts = 50
+	)
+
+	// callStartStop starts the neutrino client, requires no error on
+	// startup, immediately stops the client and waits for shutdown.
+	// The returned channel is closed once shutdown is complete.
+	callStartStop := func() <-chan struct{} {
+		done := make(chan struct{})
+
+		go func() {
+			defer close(done)
+
+			err := nc.Start()
+			require.NoError(t, err)
+			nc.Stop()
+			nc.WaitForShutdown()
+		}()
+
+		return done
+	}
+
+	// For each wanted restart, execute callStartStop and wait until the
+	// call is done before continuing to the next execution.  Waiting for
+	// a read from done forces all executions of callStartStop to be done
+	// sequentially.
+	//
+	// The test fails if all of the wanted restarts cannot be completed
+	// sequentially before the timeout is reached.
+	timeout := time.After(maxDur)
+	for i := 0; i < wantRestarts; i++ {
+		select {
+		case <-timeout:
+			t.Fatal("timed out")
+		case <-callStartStop():
+		}
+	}
+}
+
+// TestNeutrinoClientNotifyReceived verifies that a call to NotifyReceived sets
+// the client into the scanning state and that subsequent calls while scanning
+// will call Update on the client's Rescanner.
+func TestNeutrinoClientNotifyReceived(t *testing.T) {
+	var (
+		nc                      = newMockNeutrinoClient()
+		wantNotifyReceivedCalls = 50
+		wantUpdateCalls         = wantNotifyReceivedCalls - 1
+	)
+
+	// executeCalls calls NotifyReceived() synchronously n times without
+	// blocking the test and requires no error after each call.
+	executeCalls := func(n int) <-chan struct{} {
+		done := make(chan struct{})
+
+		go func() {
+			defer close(done)
+
+			var addrs []btcutil.Address
+			for i := 0; i < n; i++ {
+				err := nc.NotifyReceived(addrs)
+				require.NoError(t, err)
+			}
+		}()
+
+		return done
+	}
+
+	// Wait for all calls to complete or test to time out.
+	timeout := time.After(maxDur)
+	select {
+	case <-timeout:
+		t.Fatal("timed out")
+	case <-executeCalls(wantNotifyReceivedCalls):
+		// Require that the expected number of calls to Update were made
+		// once done sending all NotifyReceived calls.
+		mockRescan := nc.rescan.(*mockRescanner)
+		gotUpdateCalls := mockRescan.updateArgs.Len()
+		require.Equal(t, wantUpdateCalls, gotUpdateCalls)
+	}
+}
+
+// TestNeutrinoClientNotifyReceivedRescan verifies concurrent calls to
+// NotifyBlocks, NotifyReceived and Rescan do not result in a data race
+// and that there is no panic on replacing the rescan goroutine single instance.
+//
+// Each successful method call writes a success message to a buffered channel.
+// The channel is buffered so that no concurrent reader is needed.  The buffer
+// size is exactly the number of goroutines launched because each goroutine
+// must finish successfully or else this test will fail.  Each message is read
+// out of the channel to verify the number of messages received is the number
+// expected (i.e., wantMsgs == gotMsgs).
+func TestNeutrinoClientNotifyReceivedRescan(t *testing.T) {
+	var (
+		addrs     []btcutil.Address
+		nc        = newMockNeutrinoClient()
+		wantMsgs  = 100
+		gotMsgs   = 0
+		msgCh     = make(chan string, wantMsgs)
+		msgPrefix = "successfully called"
+
+		// sendMsg writes a message to the buffered message channel.
+		sendMsg = func(s string) {
+			msgCh <- fmt.Sprintf("%s %s", msgPrefix, s)
+		}
+	)
+
+	// Define closures to wrap desired neutrino client method calls.
+
+	// cleanup is the shared cleanup function for a closure executing
+	// a neutrino client method call.  It sends a message and then
+	// decrements the wait group counter.
+	cleanup := func(wg *sync.WaitGroup, s string) {
+		defer wg.Done()
+		sendMsg(s)
+	}
+
+	// callRescan calls the Rescan() method and asserts it completes
+	// with no errors. Rescan() is called with the hash of an empty header
+	// on each call.
+	startHash := new(wire.BlockHeader).BlockHash()
+	callRescan := func(wg *sync.WaitGroup) {
+		defer cleanup(wg, "rescan")
+
+		err := nc.Rescan(&startHash, addrs, nil)
+		require.NoError(t, err)
+	}
+
+	// callNotifyReceived calls the NotifyReceived() method and asserts it
+	// completes with no errors.
+	callNotifyReceived := func(wg *sync.WaitGroup) {
+		defer cleanup(wg, "notify received")
+
+		err := nc.NotifyReceived(addrs)
+		require.NoError(t, err)
+	}
+
+	// callNotifyBlocks calls the NotifyBlocks() method and asserts it
+	// completes with no errors.
+	callNotifyBlocks := func(wg *sync.WaitGroup) {
+		defer cleanup(wg, "notify blocks")
+
+		err := nc.NotifyBlocks()
+		require.NoError(t, err)
+	}
+
+	// executeCalls launches the wanted number of goroutines, waits
+	// for them to finish and signals all done by closing the returned
+	// channel.
+	executeCalls := func(n int) <-chan struct{} {
+		done := make(chan struct{})
+
+		go func() {
+			defer close(done)
+
+			var wg sync.WaitGroup
+			defer wg.Wait()
+
+			wg.Add(n)
+			for i := 0; i < n; i++ {
+				if i%3 == 0 {
+					go callRescan(&wg)
+					continue
+				}
+
+				if i%10 == 0 {
+					go callNotifyBlocks(&wg)
+					continue
+				}
+
+				go callNotifyReceived(&wg)
+			}
+		}()
+
+		return done
+	}
+
+	// Start the client.
+	err := nc.Start()
+	require.NoError(t, err)
+
+	// Wait for all calls to complete or test to time out.
+	timeout := time.After(maxDur)
+	select {
+	case <-timeout:
+		t.Fatal("timed out")
+	case <-executeCalls(wantMsgs):
+		// Ensure that exactly wantRoutines number of calls were made
+		// by counting the results on the message channel.
+		close(msgCh)
+		for str := range msgCh {
+			assert.Contains(t, str, msgPrefix)
+			gotMsgs++
+		}
+
+		require.Equal(t, wantMsgs, gotMsgs)
+	}
+}

--- a/chain/rescan.go
+++ b/chain/rescan.go
@@ -1,0 +1,39 @@
+package chain
+
+import "github.com/lightninglabs/neutrino"
+
+var _ rescanner = (*neutrino.Rescan)(nil)
+
+// rescanner is an interface that abstractly defines the public methods of
+// a *neutrino.Rescan.  The interface is private because it is only ever
+// intended to be implemented by a *neutrino.Rescan.
+type rescanner interface {
+	starter
+	updater
+
+	// WaitForShutdown blocks until the underlying rescan object is shutdown.
+	// Close the quit channel before calling WaitForShutdown.
+	WaitForShutdown()
+}
+
+// updater is the interface that wraps the Update method of a rescan object.
+type updater interface {
+	// Update targets a long-running rescan/notification client with
+	// updateable filters.  Attempts to update the filters will fail
+	// if either the rescan is no longer running or the shutdown signal is
+	// received prior to sending the update.
+	Update(...neutrino.UpdateOption) error
+}
+
+// starter is the interface that wraps the Start method of a rescan object.
+type starter interface {
+	// Start initializes the rescan goroutine, which will begin to scan the chain
+	// according to the specified rescan options.  Start returns a channel that
+	// communicates any startup errors.  Attempts to start a running rescan
+	// goroutine will error.
+	Start() <-chan error
+}
+
+// newRescanFunc defines a constructor that accepts rescan options and returns
+// an object that satisfies rescanner interface.
+type newRescanFunc func(...neutrino.RescanOption) rescanner

